### ISSUE TITLE
"awaiting reply" and "changes requested" labels expiring time

### ISSUE THREAD

**Issue by mondeja** [simple-icons/simple-icons#4461]
I think that we should discuss how long can be open a pull or issue if they are marked with these labels because there are a lot of open pull requests that could be abandoned and are getting in the way of updating or adding icons and a lot of issues potentially invalid awaiting for response.

- [ ] Pull request with `awaiting reply` without response.
- [ ] Pull request with `changes requested` without changes.
- [ ] Issue with `awaiting reply` without response from the user that opened it.
- [ ] Issue with `awaiting reply` without third party response.

**Comment by fbernhart** [comment]
I think for now we're still having more than enough work with #3169 and other open PRs. So we could give them some more time (I guess that especially during the holiday season people maybe won't reply/work on their PRs.)

But generally speaking I would say maybe a month?

**Comment by adamrusted** [comment]
I was thinking the same for the "Good First Issue" posts - once we've worked through the backlog, do we need to define how long those are left open for new contributors before an existing contributor picks it up? Since my first icon was approved, I've been trying to steer clear of them, but I'm aware that means some projects are just sat for ages waiting for someone.

**Comment by fbernhart** [comment]
@adamrusted  There was a comment once, that "good first issues" are labeled as such for 7 days. https://github.com/simple-icons/simple-icons/issues/3224#issuecomment-647481158

So I guess feel free to work on any icon request that is older than 7 days. Probably it would be the best to start with the oldest ones, that are already several months or weeks old.

**Comment by mondeja** [comment]
> I was thinking the same for the "Good First Issue" posts

Good point :+1:

> So I guess feel free to work on any icon request that is older than 7 days.

This margin seems reasonable to me too.

> I think for now we're still having more than enough work with #3169 and other open PRs.

Yes! For now, we should end the 2020 review and in 2021 we will define adequate margins for label expiration, taking into account the work we can carry out.

**Comment by NovaGL** [comment]
Once we sort the back log we can look at the average time for an issue.

Good first issue mainly means the logo should be easily added. I don't think PRs should be closed, I think someone should try and fix them first.

But I would make an exception if an issue is waiting on a reply as not enough information is provided eg not filled in I would give it a month then close it.

**Comment by adamrusted** [comment]
> Good first issue mainly means the logo should be easily added. I don't think PRs should be closed, I think someone should try and fix them first.

I meant more so on the issue front as opposed to PRs, obviously we should be supporting new contributors if they've gotten to the point of submitting a PR 😅

A stale bot may come in handy, I've started bumping a few idle PRs, but a warning from a bot that it's classed as `stale` and will close soon may give people a nudge.

**Comment by mondeja** [comment]
To clarify this a bit, not only the expiring time, as @adamrusted has been doing with some old potentially abandoned pull requests, those with `awaiting reply` and `changes requested` labels that would exceed their expiring time should be?

- Fixed by maintainers if possible, which has push permissions to contributors branches (@mondeja votes for this approach).
- Marked with `abandoned` label and closed.
- Marked with `abandoned` label and keep opened until a fix is pushed by maintainers.
- Other approach (please, suggest it).

**Comment by adamrusted** [comment]
> * Fixed by maintainers if possible, which has push permissions to contributors branches (@mondeja votes for this approach).

I also think this is the best approach. If the PRs are marked as 'stale', we can then sort through those and try and either make contact with the contributor or fix issues ourselves.

> * Marked with `abandoned` label and closed.

Some of the ones I've looked at today would have been `stale` for silly reasons, like the linter failing on an lower-case hex code, or a path not being centred. It makes sense for us to try and fix this before closing the PR and reopening a new one.

> * Marked with `abandoned` label and keep opened until a fix is pushed by maintainers.

This could also work, though one or two I've encountered today no longer have the repo for me to clone from.

**Comment by ericcornelissen** [comment]
Some thoughts, without replying to any comments specifically:

- Automatically labelling issues or PRs stale is a good idea. It may come across as unpersonal to outsiders, but unfortunately we have limited time to keep track of everything. I think marking stale should not close issues or PRs for the following reason: it is very discouraging for contributors if something is closed because we didn't have time to look at it yet, stale is slightly less worse and just an opportunity for us/the contributor to see if anything still needs to happen. (Closing the issue/PR after it as been marked stale would be okay, but such that we are (nearly) always able to respond in time if needed.)
- `Good first issue`s should be left untouched by @simple-icons/maintainers for some time to allow outsiders to contribute, whether this outsider has contributed before or not is irrelevant. (<sup>I don't think this is controversial, but some of the writing suggests that existing outside contributors should not pick up such issues.</sup>)
- As @PeterShaggyNoble has suggested elsewhere, when the `awaiting reply` label is there because we're waiting for a third-party, we should definitely just go with the best we can after some time. More often than not the potential chance because of the input from the third-party will be backwards compatible in which case it is easy to change later. Potentially we should consider a longer window if we expect the feedback of the third party would not be backwards compatible.
- PRs that seem to have been abandoned can always be picked up by maintainers. But, depending on how long it has been since the last activity I think we should ask in the PR whether the original contributor still intents to complete the process (and then we need a set amount of time before that can be asked again/is ignored).
- The `abandoned` label should only be applied to PRs that are not merged because they were not finished. I think `stale` is the better option for PRs that are still open but seem abandoned.

**Comment by PeterShaggyNoble** [comment]
To formalise some of the suggestions I made internally.

## Preamble
Our goal as maintainers should be to encourage outside contributions while, at the same time, not leaving requests sitting idle for too long; we don't _need_ to take everything on ourselves. (Of course, we're all here initially because of love of creating icons so nobody's going to fault anyone for taking too much on board!) So I would firstly suggest that we allow for a little leeway in these guidelines, at the discretion of each individual team member, particularly in the times we're living in when people's lives are in chaos - I can personally attest to not having the mental energy for icon work for a good few months.

For each label/situation here's what I suggest:

## "Stale" Issues & PRs
It would be great if we could get a bit up & running for these but instead of naming the label "stale", we should go with something less negative like "requires attention". It should also, if possible, leave a comment when adding the label so it gets bumped into our notifications. I'll include my suggestions for what action the bot should take for each label below, which may necessitate the creation of a few new labels.

## Good First Issue
Up until now, we've been leaving these for at least 7 days but, at the time that decision was made, we had a smaller, less active team so it was common for these issues to last longer than 7 days. Now, though, I would suggest extending that to 14 days to give outside contributors more of a chance to pick them up and, although they are usually the same, that should be 14 days from the label being added rather than the request being made so that the issue is listed on [Up For Grabs](https://up-for-grabs.net/) for 2 weeks.

**Stale-bot action:** No action required

## Awaiting Reply (Incomplete issue template)
Should be closed 7 days after being asked to provide the required information if it hasn't been.

**Stale-bot action:** Close issue with a canned response after 7 days

## Awaiting Reply (... from third party)
If there is no response after 14 days and the information we were waiting on is not directly blocking us from doing so then we proceed as best we can (e.g., clarification of brand colour). If it is blocking us from proceeding then we wait another 14 days (28 total) and then decide what to do on a case-by-case basis. If we've pinged the third party for a response, I don't think we should hassle them by pinging them again but, if we've created an issue in their repo then we could bump it either after 14 or 28 days.

**Stale-bot action:** Comment & add label after 14 days and comment again after a further 14 days

## Awaiting Reply (All other cases)
We ping the person we're awaiting a response from after 14 days. If they respond without action within those 14 days then we give another 14 days before pinging them again. If after another 14 days (from the first or second ping) there is still no action then we proceed without them with whatever action is needed.

**Stale-bot action:** Comment & add label after 14 days and comment again every 14 days

## Changes Requested (Minor changes)
If minor changes have been requested, which would usually be something in the JSON file, then, in the interests of getting the PR merged, we push the necessary changes ourselves after 7 days. If the contributor hasn't set their PR up to allow us to push the changes we instead ping them to remind them after the 7 days and, if they haven't responded in another 7 days, we close the PR and submit a new one, pinging the original contributor to credit them.

**Stale-bot action:** Comment & add label after 7 days and comment again after a further 7 days

## Changes Requested (Major changes)
If major changes have been requested, which would usually be an issue with the SVG, then we follow the same actions as when we're awaiting a reply in all other cases and if no action has been taken 14 days after the last ping we close the PR and label it "abandoned". If the contributor was working on someone else's request then we can either post a comment in the original issue inviting someone else to contribute or pick up the PR ourselves. If the contributor was working on their own request, I see no harm in leaving it at that until someone else requests the same icon.

**Stale-bot action:** Comment & add label after 14 days and comment again every 14 days

## All Other Cases
In the majority of all other cases, issues & PRs will be sitting idle simply because they've fallen between the cracks and it's actually a response or action from _us_ that's needed.

**Stale-bot action:** Comment & add label after 14 days and comment again every 14 days

**Comment by ericcornelissen** [comment]
Great write up @PeterShaggyNoble, I agree with pretty much everything, I'm just not sure about the repeat comments every X days 🤔 Anyway, my only worry is that existing stale bots are not build for this kind of "complexity" 😅 

> It would be great if we could get a bit up & running for these but instead of naming the label "stale", we should go with something less negative like "requires attention". 

I think this is truly one of the most important points! In the same spirit as using `out of scope` instead of `invalid` 👍

**Comment by PeterShaggyNoble** [comment]
> Anyway, my only worry is that existing stale bots are not build for this kind of "complexity"

I have no knowledge of GH bots so I'll leave the particulars to those that do 😆

> In the same spirit as using `out of scope` instead of `invalid`

We could probably retire the `invalid` label as it doesn't serve much purpose. If an issue/PR is completely off-topic for us then we can just remove all labels when closing it and, possibly, lock it as well.

**Comment by mondeja** [comment]
> We could probably retire the invalid label as it doesn't serve much purpose. If an issue/PR is completely off-topic for us then we can just remove all labels when closing it and, possibly, lock it as well.

I think the same, but why not add `off-topic` to avoid the need of enter in the issue/PR to check the closing reason?

**Comment by fbernhart** [comment]
> We could probably retire the `invalid` label as it doesn't serve much purpose.

As far as I know the `invalid` label is needed for Hacktoberfest. By looking at past PRs that had been labeled with "invalid", it seems like Hacktoberfest hasn't been really a problem to our repo.

But I would still keep it for the future.

**Comment by ericcornelissen** [comment]
@jorgeamadosoria, you might be interested in this as it is tangentially related to the discussion in #4629

**Comment by jorgeamadosoria** [comment]
> @jorgeamadosoria, you might be interested in this as it is tangentially related to the discussion in #4629

I am, thanks @ericcornelissen.
- I agree with @PeterShaggyNoble's writeup, except that I would extend some intervals to 1 month. 
For example, I would leave "good first issue" up for 1 month, then remove the label so anyone can take it.

- I guess we would have to write a stalebot for this, since I agree with @ericcornelissen that most existing ones are not going to be able to handle this specificity.

- I agree with @mondeja in that we should label PRs with off-topic rather than removing all labels. That would avoid constantly checking closed PRs and issues because they are not neatly labelled as they should.

I find this very useful in general.

**Comment by service-paradis** [comment]
Now that we have Issue forms and we ask people if they want to contribute, I think we might extend the "good first issue" rule. If we dont add this label, but the requester answer Yes, we should also wait some time before taking it. #6592 is an example where we could have waited a little bit more (not saying it was not correct to take it) since the interest of contributing was mentioned.

Also, should we document those ["rules"](https://github.com/simple-icons/simple-icons/issues/4461#issuecomment-759387848) somewhere more official?

**Comment by ericcornelissen** [comment]
> Now that we have Issue forms and we ask people if they want to contribute, I think we might extend the "good first issue" rule. If we dont add this label, but the requester answer Yes, we should also wait some time before taking it. #6592 is an example where we could have waited a little bit more (not saying it was not correct to take it) since the interest of contributing was mentioned.

I personally (still? not sure if I voiced this before) don't see a reason for remove the `good first issue`, I would rather say that we as maintainers wait x time before picking those up. 


> Also, should we document those "rules" somewhere more official?

Definitely - I think that would be the last step before closing this issue (automation can be taken care of separately, the issue says "define" after all). My vote would go for a `LABELS.md` file so it's easier to link to and so the `CONTRIBUTING.md` file doesn't grow even bigger (following #5628 and #6023).

**Comment by uncenter** [comment]
I stumbled upon this issue earlier today and I think it is still relevant seeing as we recently introduced new guidelines regarding the https://github.com/simple-icons/simple-icons/labels/permission%20required label (TLDR; 2 weeks once marked as https://github.com/simple-icons/simple-icons/labels/permission%20required for anyone but preferably the requester to reach out to the brand, if contact is made a further 2 months will be allotted to wait for a response from the brand).

I wasn't involved with this issue when the primary discussion occurred but I've given it a skim and I think the major points are (summarizing @PeterShaggyNoble's summary 😅):

1. Bumping inactive issues/PRs as to not neglect requesters or contributors
2. How long https://github.com/simple-icons/simple-icons/labels/good%20first%20issue issues should be reserved for non-contributors for
3. How long https://github.com/simple-icons/simple-icons/labels/incomplete issues should be left open for
4. Awaiting clarification reply from third party (e.g. brand color, preferred icon/variation)
5. Awaiting reply from a contributor in a PR
6. Changes requested on a PR

1 - I think a "stale" bot/workflow does more harm than good. As maintainers we want to certainly not neglect older issues and PRs, though needlessly pinging the author does more harm than good - I've been going with a strategy of a carefully refined GitHub search filter: 
```
is:issue is:open sort:updated-asc -linked:pr  -label:"permission required" -label:assessing -label:pending -label:meta -label:package -label:"good first issue"
```
This searches for open issues, starting with the ones updated the longest ago, and not including any with linked PRs or with any of the following labels: https://github.com/simple-icons/simple-icons/labels/permission%20required, https://github.com/simple-icons/simple-icons/labels/assessing, https://github.com/simple-icons/simple-icons/labels/pending, https://github.com/simple-icons/simple-icons/labels/meta, https://github.com/simple-icons/simple-icons/labels/package, or https://github.com/simple-icons/simple-icons/labels/good%20first%20issue.

2 - I don't think having a hard requirement here is necessary, though a good rule of thumb is to always have a couple open to keep the repository inviting to new contributors. Currently we have 30 good first issues, so a few done by a contributor here and there won't hurt. Again, the filter I just mentioned has 300+ issues found so regular contributors should hopefully be aiming to knock those out instead of the good first issues anyway.

3 - Pretty quickly I would think, I liked the timeframe of 7 days.
4 - Sticking with our 2 month allotment as we do for https://github.com/simple-icons/simple-icons/labels/permission%20required seems reasonable.

5 & 6 - I don't think this should be a hard and fast rule either, but if a contributor doesn't respond to a request for changes and a follow up comment within a few months its safe to mark as https://github.com/simple-icons/simple-icons/labels/abandoned and close it.

**Comment by adamrusted** [comment]
I agree with @uncenter's points above.

1. We should look to nudge contributors toward our oldest outstanding issues first via a filter similar to the one listed above.
2. As a general rule of thumb, I personally will still try and keep away from https://github.com/simple-icons/simple-icons/labels/good%20first%20issue issues unless it's a brand I'm desperate to see included.
3. Timeframe of 7 days from submission of issue is a good time to close out https://github.com/simple-icons/simple-icons/labels/incomplete issues - providing us maintainers are quick to the draw with highlighting missing info. If it's been sat for some time, we should make some allowances for authors to revisit (perhaps the 7 days mentioned above, or as with our https://github.com/simple-icons/simple-icons/labels/permission%20required initial response window of 2 weeks).
4. Agree on third-party responses. If a contributor to an issue or PR can demonstrate they've reached out to the third party, we should give the third party 2 months to respond. For an OSS project with active maintainers, that's likely to be massively overkill - but for a lot of brands with legal departments and a lot of bureaucracy this may just be enough.
5. Again, if we're quick to the draw with responding, I don't think it's unreasonable to expect a response from a contributor within 2 weeks. If it's been sat for a while, I can see the reluctance to return to it.
6. On this one, my personal rule of thumb has been to:
  - Ask if the contributor would still like to work on it
  - Wait 2 weeks
  - Mark as abandoned, but keep PR open for a brief window (maybe the 2 months - for consistency sake) for maintainers to pick it up
  - Close the existing PR, unlink from the Issue so it shows back up in the search criteria of point 1.
  -

**Comment by PeterShaggyNoble** [comment]
I think this can be closed out, no? We have pretty good rules of thumb in place now, and internal tools for tracking a lot of it.

I don't think we need anything explicit in our contributing guidelines, as most all of it as an internal process. But, if were were going to add it anywhere, then I would suggest it be done as part of revisiting #6622.
